code/headers/microphone_controller.hpp

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include "hwlib.hpp"
+
+namespace r2d2::microphone {
+
+    /**
+     * Class microphone_controller_c provides functions
+     * for collecting audio data from the KY-038 module.
+     *
+     * Uses a template to determine the buffer size at
+     * compile time.
+     */
+
+    class microphone_controller_c {
+    private:
+        hwlib::adc &mic;
+	static const int buffer_size = 63; // size of internal communication buffer, this size must not be changed!!
+	uint16_t sample_buffer[buffer_size];
```

Comment:
Instead of a C-style array, I would like to see a `std:array` here. That way, you `read_buffer` function could return a reference to the array instead of a pointer to the first element. This helps, because then you aren't depending on the internal size of the buffer in order to navigate through the buffer.